File: packages/cortex-js/tests/resources/pools.test.ts
import { describe, it, expect } from 'vitest';
import { ValidationError } from '../../src/errors.js';
import { mockFetch, createClient } from '../helpers.js';

describe('pools', () => {
  describe('list', () => {
    it('lists all pools', async () => {
      const mockResponse = {
        pools: [{ id: 'pool-1', name: 'Default', createdAt: '2024-01-01', updatedAt: '2024-01-01' }],
      };
      const fetch = mockFetch(mockResponse);
      const client = createClient(fetch);

      const result = await client.pools.list();

      expect(result.pools).toHaveLength(1);
      expect(result.pools[0]!.name).toBe('Default');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/pools');
      expect(opts.method).toBe('GET');
    });
  });

  describe('create', () => {
    it('creates a pool', async () => {
      const mockResponse = { id: 'pool-2', name: 'New Pool', createdAt: '2024-01-01', updatedAt: '2024-01-01' };
      const fetch = mockFetch(mockResponse);
      const client = createClient(fetch);

      const result = await client.pools.create({ name: 'New Pool' });

      expect(result.id).toBe('pool-2');
      expect(result.name).toBe('New Pool');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/pools');
      expect(opts.method).toBe('POST');
      expect(JSON.parse(opts.body as string)).toEqual({ name: 'New Pool' });
    });

    it('throws ValidationError when name is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.pools.create({ name: '' })).rejects.toThrow(ValidationError);
    });
  });

  describe('update', () => {
    it('updates a pool', async () => {
      const fetch = mockFetch({ id: 'pool-1', name: 'Updated', createdAt: '2024-01-01', updatedAt: '2024-01-02' });
      const client = createClient(fetch);

      const result = await client.pools.update('pool-1', { name: 'Updated' });

      expect(result.name).toBe('Updated');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/pools/pool-1');
      expect(opts.method).toBe('PATCH');
    });

    it('throws ValidationError when id is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.pools.update('', { name: 'x' })).rejects.toThrow(ValidationError);
    });
  });

  describe('delete', () => {
    it('deletes a pool', async () => {
      const fetch = mockFetch({ id: 'pool-1', deleted: true });
      const client = createClient(fetch);

      const result = await client.pools.delete('pool-1');

      expect(result.deleted).toBe(true);

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/pools/pool-1');
      expect(opts.method).toBe('DELETE');
    });

    it('throws ValidationError when id is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.pools.delete('')).rejects.toThrow(ValidationError);
    });
  });

  describe('addBackend', () => {
    it('adds a backend to a pool', async () => {
      const fetch = mockFetch({ id: 'pb-1', backendId: 'b-1', priority: 1 });
      const client = createClient(fetch);

      const result = await client.pools.addBackend('pool-1', { backendId: 'b-1', priority: 1 });

      expect(result.backendId).toBe('b-1');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/pools/pool-1/backends');
      expect(opts.method).toBe('POST');
    });

    it('throws ValidationError when poolId is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.pools.addBackend('', { backendId: 'b-1' })).rejects.toThrow(ValidationError);
    });

    it('throws ValidationError when backendId is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.pools.addBackend('pool-1', { backendId: '' })).rejects.toThrow(ValidationError);
    });
  });

  describe('removeBackend', () => {
    it('removes a backend from a pool', async () => {
      const fetch = mockFetch({ id: 'b-1', deleted: true });
      const client = createClient(fetch);

      const result = await client.pools.removeBackend('pool-1', 'b-1');

      expect(result.deleted).toBe(true);

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/pools/pool-1/backends/b-1');
      expect(opts.method).toBe('DELETE');
    });

    it('throws ValidationError when poolId is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.pools.removeBackend('', 'b-1')).rejects.toThrow(ValidationError);
    });

    it('throws ValidationError when backendId is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.pools.removeBackend('pool-1', '')).rejects.toThrow(ValidationError);
    });
  });
});

File: packages/cortex-dart/test/resources/pools_test.dart
import 'dart:convert';

import 'package:nfinitmonkeys_cortex_sdk/cortex_sdk.dart';
import 'package:http/http.dart' as http;
import 'package:test/test.dart';

import '../helpers.dart';

Map<String, dynamic> samplePool({
  String id = 'pool-123',
  String name = 'Default Pool',
}) =>
    {
      'id': id,
      'name': name,
      'description': 'A test pool',
      'backends': ['backend-1', 'backend-2'],
      'createdAt': '2024-01-01T00:00:00Z',
    };

void main() {
  group('PoolsResource', () {
    late CortexClient client;

    tearDown(() => client.close());

    test('list returns pools', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {
          'data': [
            samplePool(id: 'pool-1', name: 'Pool A'),
            samplePool(id: 'pool-2', name: 'Pool B'),
          ],
        }),
      );

      final pools = await client.pools.list();
      expect(pools, hasLength(2));
      expect(pools.first.id, 'pool-1');
      expect(pools.last.name, 'Pool B');
    });

    test('list handles empty response', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {'data': <dynamic>[]}),
      );

      final pools = await client.pools.list();
      expect(pools, isEmpty);
    });

    test('create sends correct body', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, samplePool(), requests),
      );

      final pool = await client.pools.create(
        name: 'My Pool',
        description: 'A new pool',
      );

      expect(pool.id, 'pool-123');
      expect(pool.name, 'Default Pool');

      final body = jsonDecode(requests.first.body) as Map<String, dynamic>;
      expect(body['name'], 'My Pool');
      expect(body['description'], 'A new pool');
    });

    test('create validates empty name', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.pools.create(name: ''),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('update sends PATCH request', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, samplePool(), requests),
      );

      final pool = await client.pools.update(
        id: 'pool-123',
        name: 'Updated Pool',
      );

      expect(pool.id, 'pool-123');
      expect(requests.first.method, 'PATCH');
      expect(requests.first.url.path, contains('/admin/pools/pool-123'));
    });

    test('update validates empty id', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.pools.update(id: ''),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('delete sends DELETE request', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, {}, requests),
      );

      await client.pools.delete('pool-123');

      expect(requests.first.method, 'DELETE');
      expect(requests.first.url.path, contains('/admin/pools/pool-123'));
    });

    test('delete validates empty id', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.pools.delete(''),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('addBackend sends correct body', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, {'ok': true}, requests),
      );

      await client.pools.addBackend(
        poolId: 'pool-123',
        backendId: 'backend-456',
      );

      expect(requests.first.method, 'POST');
      expect(
        requests.first.url.path,
        contains('/admin/pools/pool-123/backends'),
      );
      final body = jsonDecode(requests.first.body) as Map<String, dynamic>;
      expect(body['backendId'], 'backend-456');
    });

    test('addBackend validates empty poolId', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.pools.addBackend(poolId: '', backendId: 'b-1'),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('removeBackend sends DELETE request', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, {}, requests),
      );

      await client.pools.removeBackend(
        poolId: 'pool-123',
        backendId: 'backend-456',
      );

      expect(requests.first.method, 'DELETE');
      expect(
        requests.first.url.path,
        contains('/admin/pools/pool-123/backends/backend-456'),
      );
    });

    test('handles 401 errors', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        maxRetries: 0,
        httpClient: mockClient(401, 'Unauthorized'),
      );

      expect(
        () => client.pools.list(),
        throwsA(isA<CortexAuthenticationException>()),
      );
    });
  });
}
